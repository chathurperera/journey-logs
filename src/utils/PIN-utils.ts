import moment from 'moment';

export const isPinSessionExpired = (lastSessionTimestamp: number, maxDurationInMinutes: number = 5) => {
  const timeDifference = moment().diff(moment(lastSessionTimestamp * 1000), 'minute');
  console.log('timeDifference', timeDifference);

  return timeDifference > maxDurationInMinutes;
};
