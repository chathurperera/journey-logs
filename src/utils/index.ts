export * from './moment-utils';
export * from './render-utils';
export * from './PIN-utils';
