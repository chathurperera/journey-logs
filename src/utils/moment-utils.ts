import moment from 'moment';

export const getCurrentTimestampInMilliSeconds = () => {
  const timestamp = moment().unix();
  return timestamp;
};

export const getRelativeTimeFromTimestamp = (timestampValue: number) => {
  const relativeTime = moment.unix(timestampValue).fromNow();
  return relativeTime;
};
