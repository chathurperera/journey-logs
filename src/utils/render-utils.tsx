import { NavigationContext } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';

import { store } from '@jl/stores';

type RenderParams = Parameters<typeof render>;

export function renderWithProviders(ui: RenderParams[0], options?: RenderParams[1]) {
  const BaseProviders = ({ children }: { children: ReactNode }) => (
    <Provider store={store}>
      <NavigationContext.Provider
        //@ts-ignore
        value={{
          isFocused: () => true,
          addListener: jest.fn(() => jest.fn()),
        }}>
        {children}
      </NavigationContext.Provider>
    </Provider>
  );

  return render(ui, { wrapper: BaseProviders, ...options });
}
