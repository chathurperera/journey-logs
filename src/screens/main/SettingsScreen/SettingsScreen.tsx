import React from 'react';
import { View } from 'react-native';

import { Text } from '@jl/components';
import { tw } from '@jl/config';
import { Color, Route, TextAlignment, TextVariant } from '@jl/constants';
import { NavigationService } from '@jl/services';
import { useDispatch, useSelector } from '@jl/stores';
import { getRemainingLockoutTime, isPinSessionExpired, validateLockoutPeriod } from '@jl/utils';

import { BaseScreenLayout } from '../../components/BaseScreenLayout';
import { Section } from './components/Section';
import { SectionLink } from './components/SectionLink';

interface SettingsScreenProps {
  testID: string;
}

export function SettingsScreen({ testID }: SettingsScreenProps) {
  const dispatch = useDispatch();
  const { securityPreference } = useSelector(state => state.encryptionStore);

  const { salt, lastAccessedHiddenNotesAt, lockoutTimestamp } = useSelector(state => state.encryptionStore);
  const { name, email } = useSelector(state => state.userStore.userData);

  const isInLockedPeriod = validateLockoutPeriod(lockoutTimestamp, securityPreference);

  const hiddenNotesAccessNavigation = () => {
    if (isInLockedPeriod) {
      const remainingSeconds = getRemainingLockoutTime(lockoutTimestamp);
      return NavigationService.navigate(Route.MaxPinCodeAttemptsReached, { remainingSeconds: remainingSeconds });
    }

    if (lastAccessedHiddenNotesAt && isPinSessionExpired(lastAccessedHiddenNotesAt)) {
      return NavigationService.navigate(Route.PinCode, { pinExists: true });
    }

    if (lastAccessedHiddenNotesAt) {
      return NavigationService.navigate(Route.HiddenNotes);
    }

    NavigationService.navigate(Route.PinCode, { pinExists: Boolean(salt) });
  };

  const handleAddPINNavigation = () => {
    if (salt === '') {
      NavigationService.navigate(Route.PinCode, { pinExists: salt !== '' });
    } else {
      NavigationService.navigate(Route.OldPINVerification);
    }
  };

  return (
    <BaseScreenLayout testID={testID}>
      <View style={tw`mx-5 flex-1`}>
        <Text variant={TextVariant.Title2} textAlign={TextAlignment.Center}>
          Settings
        </Text>
        <View style={tw`pt-7 items-center`}>
          <View style={tw`w-18 h-18  bg-[${Color.Primary.Jl500}] justify-center rounded-full flex-row pt-1 mb-3`}>
            <Text
              variant={TextVariant.Heading1Regular}
              // textAlign={TextAlignment.Center}
              color={Color.Neutral.white}>
              {name?.charAt(0)}
            </Text>
          </View>
          <Text variant={TextVariant.Body1Regular} color={Color.Neutral.JL400}>
            {email}
          </Text>
        </View>
        <View style={tw`mt-10`}>
          <Section title="account">
            <SectionLink text="Account" onPress={() => NavigationService.navigate(Route.Account)} />
            <SectionLink text="Change Password" onPress={() => NavigationService.navigate(Route.ChangePassword)} />
          </Section>
          <Section title={'notes'}>
            <SectionLink text="Tags" onPress={() => NavigationService.navigate(Route.Tags)} />
            <SectionLink text={salt === '' ? 'Add PIN' : 'Change PIN'} onPress={handleAddPINNavigation} />
            {salt !== '' && <SectionLink text="Hidden Notes" onPress={() => hiddenNotesAccessNavigation()} />}
          </Section>
          <SectionLink text="Logout" onPress={() => dispatch.userStore.logoutUser()} />
        </View>
      </View>
    </BaseScreenLayout>
  );
}
