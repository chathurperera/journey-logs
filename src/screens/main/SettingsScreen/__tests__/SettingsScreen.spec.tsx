import * as React from 'react';

import { renderWithProviders } from '@jl/utils';

import { SettingsScreen } from '../SettingsScreen';

describe('login screen', () => {
  const testID = 'SettingsScreenTestID';
  const getRenderedScreen = () => renderWithProviders(<SettingsScreen testID={testID} />);

  it('should render SettingsScreen correctly', () => {
    const renderer = getRenderedScreen();
    const renderTree = renderer.toJSON();
    expect(renderTree).toMatchSnapshot();
  });
  it('should find the SettingsScreen via testID', () => {
    const { getByTestId } = getRenderedScreen();
    const foundScreen = getByTestId(testID);
    expect(foundScreen).toBeTruthy();
  });
});
