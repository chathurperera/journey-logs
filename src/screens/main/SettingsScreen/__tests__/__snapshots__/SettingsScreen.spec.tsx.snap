// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login screen should render SettingsScreen correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#f1f1f1f5",
      "height": "100%",
      "paddingTop": 50,
      "position": "relative",
    }
  }
  testID="SettingsScreenTestID"
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  />
</RCTSafeAreaView>
`;
